File: backend/face.py
#!/usr/bin/python

from pathlib import Path
import sqlite3


class Face(object):

    def __init__(self):
        self.id = None
        self.head = ""
        self.hair = ""
        self.hairX = 0
        self.hairY = 0
        self.brows = ""
        self.browsLX = 0
        self.browsLY = 0
        self.browsRX = 0
        self.browsRY = 0
        self.eyes = ""
        self.eyesLX = 0
        self.eyesLY = 0
        self.eyesRX = 0
        self.eyesRY = 0
        self.nose = ""
        self.noseX = 0
        self.noseY = 0
        self.mouth = ""
        self.mouthX = 0
        self.mouthY = 0

    def saveToDB(self, con, cur, id):
        sql = ""
        data = None

        # this is a new entry
        if id is None:
            sql = """
                INSERT INTO Faces
                (
                head, hair, hairX, hairY, brows, browsLX, browsLY, browsRX, browsRY,
                eyes, eyesLX, eyesLY, eyesRX, eyesRY, nose, noseX, noseY, mouth, mouthX, mouthY
                )
                VALUES(?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?);
            """

            data = (
                self.head, self.hair, self.hairX, self.hairY, self.brows, self.browsLX, self.browsLY, self.browsRX,
                self.browsRY, self.eyes, self.eyesLX, self.eyesLY, self.eyesRX, self.eyesRY, self.nose, self.noseX,
                self.noseY, self.mouth, self.mouthX, self.mouthY
            )

        # this is an update
        else:
            sql = """
                UPDATE Faces SET
                head = ?, hair = ?, hairX = ?, hairY = ?, brows = ?, browsLX = ?, browsLY = ?,
                browsRX = ?, browsRY = ?, eyes = ?, eyesLX = ?, eyesLY = ?, eyesRX = ?, eyesRY = ?,
                nose = ?, noseX = ?, noseY = ?, mouth = ?, mouthX = ?, mouthY = ?,
                WHERE id = ?;
            """

            data = (
                self.head, self.hair, self.hairX, self.hairY, self.brows, self.browsLX, self.browsLY, self.browsRX,
                self.browsRY, self.eyes, self.eyesLX, self.eyesLY, self.eyesRX, self.eyesRY, self.nose, self.noseX,
                self.noseY, self.mouth, self.mouthX, self.mouthY, self.id
            )

        try:
            cur.execute(sql, data)
            con.commit()

        except con.Error as error:
            print("Failed to save face", error)

    def loadFromDBbyID(self, cur, id):
        sql = """
            SELECT * FROM faces WHERE id={};
        """.format(id)

        cur.execute(sql)
        p = cur.fetchone()

        if p:
            self.id = id
            self.head = p[1]
            self.hair = p[2]
            self.hairX = p[3]
            self.hairY = p[4]
            self.brows = p[5]
            self.browsLX = p[6]
            self.browsLY = p[7]
            self.browsRX = p[8]
            self.browsRY = p[9]
            self.eyes = p[10]
            self.eyesLX = p[11]
            self.eyesLY = p[12]
            self.eyesRX = p[13]
            self.eyesRY = p[14]
            self.nose = p[15]
            self.noseX = p[16]
            self.noseY = p[17]
            self.mouth = p[18]
            self.mouthX = p[19]
            self.mouthY = p[20]

    def print(self):
        print(self.head)
        print(self.hair)

File: draw/face.lua
-- --------------------------------------------------------
--
-- draw/face.lua
--
-- andreaszdw@googlemail.com
--
-- --------------------------------------------------------
local sqlite3 = require("sqlite3")
local class = require("30log")

local face = class()

-- --------------------------------------------------------
--
-- constructor
--
-- --------------------------------------------------------
function face:init(parent, x, y, scale)
	self.parent = parent
	self.x = x or 0
	self.y = y or 0
	self.scale = scale or 1
	self.id = id
	self.head = "tint4_head.png"
	self.hair = "brown2Man1.png"
	self.hairX = 2
	self.hairY = 49
	self.brows = "brown2Brow2.png"
	self.browsLX = -32
	self.browsLY = 34
	self.browsRX = 33
	self.browsRY = 31
	self.eyes = "eyePine_small.png"
	self.eyesLX = -34
	self.eyesLY = 9
	self.eyesRX = 31
	self.eyesRY = 14
	self.nose = "tint4Nose3.png"
	self.noseX = -3
	self.noseY = -11
	self.mouth = "mouth_teethUpper.png"
	self.mouthX = 3
	self.mouthY = -44
	self.group = display.newGroup()
	self.group.x = parent.x
	self.group.y = parent.y
	self.path = "assets\\images\\faces\\"

	self.parent.view:insert(self.group)
end

-- --------------------------------------------------------
--
-- loadFromDB
--
-- --------------------------------------------------------
function face:loadFromDB(db, id)

	local sql = [[SELECT * FROM Faces WHERE id=]]..id..[[;]]

	function readFace(udata, cols, values, names)
		self.id = values[1]
        self.head = values[2]
        self.hair = values[3]
        self.hairX = values[4]
        self.hairY = values[5]
        self.brows = values[6]
        self.browsLX = values[7]
        self.browsLY = values[8]
        self.browsRX = values[9]
        self.browsRY = values[10]
        self.eyes = values[11]
        self.eyesLX = values[12]
        self.eyesLY = values[13]
        self.eyesRX = values[14]
        self.eyesRY = values[15]
        self.nose = values[16]
        self.noseX = values[17]
        self.noseY = values[18]
        self.mouth = values[19]
        self.mouthX = values[20]
        self.mouthY = values[21]
	end

	db:exec(sql, readFace)
end

-- --------------------------------------------------------
--
-- draw
-- 
-- --------------------------------------------------------
function face:draw(x, y, scale)
	self.x = x or self.x
	self.y = y or self.y
	self.scale = scale or self.scale

	local headImage = display.newImage(self.group, self.path .. self.head)

	local hairImage = display.newImage(self.group, self.path .. self.hair)
	hairImage.x = 0 + self.hairX
	hairImage.y = 0 - self.hairY

	local browImage = self.path .. self.brows
	local leftBrowImage = display.newImage(self.group, browImage)
	leftBrowImage.x = 0 + self.browsLX
	leftBrowImage.y = 0 - self.browsLY

	local rightBrowImage = display.newImage(self.group, browImage)
	rightBrowImage:scale(-self.scale, self.scale)
	rightBrowImage.x = 0 + self.browsRX
	rightBrowImage.y = 0 - self.browsRY 

	local eyeImage = self.path .. self.eyes
	local leftEyeImage = display.newImage(self.group, eyeImage)
	leftEyeImage.x = 0 + self.eyesLX
	leftEyeImage.y = 0 - self.eyesLY

	local rightEyeImage = display.newImage(self.group, eyeImage)
	rightEyeImage.x = 0 + self.eyesRX
	rightEyeImage.y = 0 - self.eyesRY

	local noseImage = display.newImage(self.group, self.path .. self.nose)
	noseImage.x = 0 + self.noseX
	noseImage.y = 0 - self.noseY

	local mouthImage = display.newImage(self.group, self.path .. self.mouth)
	mouthImage.x = 0 + self.mouthX
	mouthImage.y = 0 - self.mouthY

	self.group:scale(self.scale, self.scale)

	self.group.x = self.x
	self.group.y = self.y
end

-- --------------------------------------------------------
return face